src/utils.rs

```diff
@@ -0,0 +1,92 @@
+//! Low-level utility functions.
+
+use core::{
+    any::{type_name, TypeId},
+    mem, ptr,
+};
+
+/// Determine if two static, generic types are equal to each other.
+#[inline(always)]
+pub(crate) fn type_eq<T: 'static, U: 'static>() -> bool {
+    // Reduce the chance of `TypeId` collisions causing a problem by also
+    // verifying the layouts match and the type names match. Since `T` and `U`
+    // are known at compile time the compiler should optimize away these extra
+    // checks anyway.
+    mem::size_of::<T>() == mem::size_of::<U>()
+        && mem::align_of::<T>() == mem::align_of::<U>()
+        && mem::needs_drop::<T>() == mem::needs_drop::<U>()
+        && TypeId::of::<T>() == TypeId::of::<U>()
+        && type_name::<T>() == type_name::<U>()
+}
+
+/// Determine if two generic types which may not be static are equal to each
+/// other.
+///
+/// This function must be used with extreme discretion, as no lifetime checking
+/// is done. Meaning, this function considers `Struct<'a>` to be equal to
+/// `Struct<'b>`, even if either `'a` or `'b` outlives the other.
+#[inline(always)]
+pub(crate) fn type_eq_non_static<T: ?Sized, U: ?Sized>() -> bool {
+    // Inline has a weird, but desirable result on this function. It can't be
+    // fully inlined everywhere since it creates a function pointer of itself.
+    // But in practice when used here, the act of taking the address will be
+    // inlined, thus avoiding a function call when comparing two types.
+    #[inline]
+    fn type_id_of<T: ?Sized>() -> usize {
+        type_id_of::<T> as usize
+    }
+
+    // What we're doing here is comparing two function pointers of the same
+    // generic function to see if they are identical. If they are not
+    // identical then `T` and `U` are not the same type.
+    //
+    // If they are equal, then they _might_ be the same type, unless an
+    // optimization step reduced two different functions to the same
+    // implementation due to having the same body. To avoid this we are using
+    // a function which references itself. This is something that LLVM cannot
+    // merge, since each monomorphized function has a reference to a different
+    // global alias.
+    type_id_of::<T>() == type_id_of::<U>()
```

Comment:
I strongly suspect this invokes undefined behaviour that makes it unsound, and that it only works by accident at present.

It assumes that (a) unused function pointers are not optimised out of existence; and (b) monomorphisation doesn’t merge these functions. I’d be surprised if the compiler guaranteed either of these. They both *really* feel like places where it only works because we don’t have a sufficiently smart compiler yet. (I know that when llogiq made his crate momo there was widespread agreement that the compiler should demonomorphise lots of stuff.)

If the compiler *does* start merging these, then you have transmute in safe code, with the type_name check bearing all the load of memory safety, which is a worrying thought.

This ain’t my crate, and I have no plan on using it (just came across it incidentally), but I suggest reverting this.

Reply:
Regarding (a): The function pointers are not unused, since the function being invoked is itself the function that returns its own function pointer. When it is optimized away (at least in my inspections of what code is generated at various optimization levels) the expression itself is also optimized away, at which point the function pointers are no longer relevant (we're left with a folded constant `true` or `false`, which is all we really wanted). I have a hard time seeing the optimization step acting much differently for this code, though I suppose anything is possible.

Regarding (b): I walked through the source of LLVM's function merging implementation at several different versions to make my "hunch" a bit more objective, and at least today this is something that LLVM won't merge. However, you are right that it is possible that this will change in the future, and this says nothing of other rustc codegen backends or of optimization steps in rustc itself.

A better summary of the trick is: "These generic functions can't be merged unless they are already merged in order to have the same address space." My intuition could very well be wrong here, but it seems like merging such a function could actually result in incorrect observable behavior in some scenarios so it isn't clear to me that an optimization like this ever _would_ be implemented intentionally, at least not without something smarter that understands the broader context of how the pointer is being used within the function.

To be sure, this is a hack, and not an ideal solution. Ideally, `TypeId` would have a way of being used without a `'static` bound with the caveats that type comparisons ignore lifetimes, since we do our own independent checking of lifetime requirements with the `LifetimeFree` trait. However, it is unlikely that this will ever be supported since RFC 1849 was retracted.

> This ain’t my crate, and I have no plan on using it (just came across it incidentally), but I suggest reverting this.

I did spend quite a bit of time weighing the pros and cons of this PR, but ultimately it did seem to me that at least currently the risk here is relatively low. Though to admit my bias, retracting this entire feature now, months after already being released, would be a pain.